cpp/src/arrow/compute/function.h

```diff
@@ -229,6 +229,12 @@ class ARROW_EXPORT Function {
 
   virtual Status Validate() const;
 
+  /// \brief Returns the pure property for this function.
+  ///
+  /// For impure functions like 'random', we should skip any simplification
+  /// for this function except it's arguments.
+  virtual bool is_impure() const { return false; }
```

Comment:
I think it's better to call this `is_pure()` and use `!is_pure()` where impurity is being checked.

Reply:
And instead of `bool is_impure = false` variables we would have `bool is_pure = true`. Less negation overall.